jscomp/runtime/caml_obj.ml

```diff
@@ -198,6 +208,17 @@ let rec caml_compare (a : Obj.t) (b : Obj.t) : int =
         else
           let len_a = Bs_obj.length a in
           let len_b = Bs_obj.length b in
+          if len_a = 0 && len_b = 0 && O.is_object a && O.is_object b then
+            begin
+              let keys_a = O.keys a in
```

Comment:
Unfortunate that this allocates though. Should we loop instead?

Reply:
we did not support `for .. in` currently, maybe it's fine to write such function in raw js?

Reply:
The thing about looping is that we have to traverse *both* objects, one item at a time, as that's the order in which comparison proceeds. I could not find a way to do that.

One could choose a different order for the comparison, e.g. as in sets sorted by keys, but again I would not know how to do that without allocating.

Reply:
Is object keys order important? This passes the type checker:
https://reasonml.github.io/en/try.html?reason=C4TwDgpgBAZg9nKBeKBvAdFARPOWBcUAlgHbAA02ARgIYBOBUAzsHaQOYC+A3AFC8AbCMCgBbEADEEhXMjRZaDQlgAWEEFko4EjAIw9+QkTRJxgaulLgyEc1NryFdWxY1XqsPIA
So you'd accidentally consider these two objs as non-equal

Reply:
It's possible to ignore the order of keys, by considering them as a sorted set. Trying that next.

Reply:
Right. So if the order of keys doesn’t matter, you can do two for-loops; check  object1 against object2, then object2 against object1. This avoids the allocations

Reply:
Discussed more offline. Seems like there's no way to avoid allocation: some ordering on keys needs to be defined, and elements compared in that order.
Two possible orders are: 1 the order in which keys appear in the objects, 2 alphabetical order.

If going for 1, one needs to compare the first key in `a` with the first key in `b`, and so on. This can't be done with a for..in.
(With a for..in one can compare the first key in `a` with the corresponding key in `b`, but this is not a well-defined order, as e.g. it's not symmetric.)

If going for 2, one needs to compare the keys in alphabetical order, and this can't be done with a for..in.

Currently choosing to go for 2, which has the property that `{x:1, y:2}` and `{y:2, x:1}` are considered equal, so insertion order does not get in the way.